KAFKA-4393: Improve invalid/negative TS handling

mjsax opened the pull request.

mjsax committed:
KAFKA-4393: Improve invalid/negative TS handling

mjsax commented:
@guozhangwang @enothereska @dguy @miguno @hjafarpour 

I am not happy with the name `RobustConsumerRecordTimestampExtractor` but could not come up with anything better... Suggestions are welcome.

I am also not sure, if we still need the TS check in `SinkNode` (and the corresponding test). I doubt it can happen that a negative TS is received. (btw: the user errors reported with exception in `KafkaProducer` are only valid for `0.10.0.0` -- in `0.10.0.1` we introduced a TS check, ie, fail-fast policy already.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+ * Retrieves built-in timestamps from Kafka messages (introduced in KIP-32: Add timestamps to Kafka message).
+ * <p>
+ * Here, "built-in" refers to the fact that compatible Kafka producer clients automatically and
+ * transparently embed such timestamps into messages they sent to Kafka, which can then be retrieved
```
Yes, we mention the KIP above, but I think we should still say sth like "the new 0.10 message format".  Why?  Because users have always been able to embed timestamps into the message keys and/or message values, but when we say "embedded timestamps" here we mean the new message field for timestamps.

mjsax replied:
I just c&p from original `ConsumerRecordTimestampExtractor` -- maybe we want to update both.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+ * transparently embed such timestamps into messages they sent to Kafka, which can then be retrieved
+ * via this timestamp extractor.
+ * <p>
+ * If <i>CreateTime</i> is used to define the built-in timestamps, using this extractor effectively provide
```
provide[s]

miguno replied:
"Define" and this section sound a bit weird.

What about sth like:

> If the embedded timestamps of messages in 0.10+ message format represent <i>CreateTime</i> (cf. Kafka broker setting `log.message.timestamp.type` and Kafka topic setting `message.timestamp.type`), this extractor effectively provides ...

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+ * Retrieves built-in timestamps from Kafka messages (introduced in KIP-32: Add timestamps to Kafka message).
+ * <p>
+ * Here, "built-in" refers to the fact that compatible Kafka producer clients automatically and
+ * transparently embed such timestamps into messages they sent to Kafka, which can then be retrieved
```
they sen[d]

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+ * <p>
+ * If you need <i>processing-time</i> semantics, use {@link WallclockTimestampExtractor}.
+ * <p>
+ * If a record has a negative timestamp value, a log message is written and the timestamp is returned.
```
Ah, I guess we need to be a bit careful here because "log" and "message" are overloaded terms in Kafka. ;-)

What about sth like:

> If a record has a negative (invalid) timestamp, the timestamp is returned as-is, but in addition a WARN message is logged in your application.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+    private static final Logger log = LoggerFactory.getLogger(RobustConsumerRecordTimestampExtractor.class);
+
+    /**
+     * Extracts the embedded meta data timestamp from the given {@link ConsumerRecord}
```
"meta data" -> "metadata"

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+     * because the record will no be processed but dropped.
+     *
+     * @param record a data record
+     * @return the embedded meta-data timestamp of the given {@link ConsumerRecord}
```
meta-data -> metadata

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+        final long timestamp = record.timestamp();
+
+        if (timestamp < 0) {
+            log.warn("Dropping input record {} because it has an invalid (negative) timestamp.", record);
```
This is IMHO quite confusing (and, ultimately, wrong).

We say "We are dropping this record", but in fact we are not -- we are returning the invalid timestamp as-is.

dguy replied:
The record gets dropped in `RecordQueue` if the timestamp is negative

miguno replied:
@dguy: I am aware that this **ultimately** happens.  My point is that the current code structure needs improvement.  `RobustConsumerRecordTimestampExtractor` should only make claims based on its local knowledge.  The above is a symptom of the problem, but not the root cause.

mjsax replied:
Will change this to `"Input record {} will be dropped because it has an invalid (negative) timestamp."`.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:
```diff
@@ -107,8 +107,9 @@ public int addRawRecords(Iterable<ConsumerRecord<byte[], byte[]>> rawRecords, Ti
             log.trace("Source node {} extracted timestamp {} for record {} when adding to buffered queue", source.name(), timestamp, record);
 
             // validate that timestamp must be non-negative
```
The code comment is no longer correct.  We are not validating any longer, we are silently skipping.

miguno reviewed (changes requested):
Right now I think I am confused about this change.

We are logging WARN messages in the new extractor (and say we are "dropping" messages with invalid timestamps, though in fact we are not doing that in the extractor -- so how can the extractor claim that to happen with its local-only knowledge?).  And in other parts of the code we are now silently dropping/skipping any such messages.

We need more clarity/structure here.  My understanding is as follows:
- The default timestamp extractor and the new timestamp extractor differ only in whether or not they are logging a WARN message when a message has an invalid timestamp.
- We are changing other parts of the code to silently skip messages with invalid timestamps.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,65 @@
+ * @see ConsumerRecordTimestampExtractor
+ * @see WallclockTimestampExtractor
+ */
+public class RobustConsumerRecordTimestampExtractor implements TimestampExtractor {
```
I agree we need a better name, and we can do that once we have more clarity on how we want to improve the current code.

miguno commented:
I think I realize that the original idea of splitting the functionalities of the timestamp extractor and  the handler (for messages with invalid timestamps) might have been the better idea:  The timestamp extractor should only be responsible for extracting the timestamp, whereas the handler (which would be used e.g. inside `addRawRecords`) should only be responsible for determining what happens to records with invalid timestamps.

At the moment, we are in the unfortunate situation where, for example, the (new) timestamp extractor is effectively identical to the default extractor with regards to timestamp extraction.  The only difference is that it also logs a WARN message (which arguably is ok -- e.g. it may log information to say "Hey, I came across a message that doesn't include a timestamp, and in my case I won't do anything about that and return a negative timestamp as-is, but at least I can let you know") saying "this message will be dropped" (and this wording/information is wrong, because, based on its local knowledge, it can't determine whether or not any downstream code will or will not drop the message).

dguy commented:
If we are going to go with this approach, i'm wondering if we should change the signature of `TimestampExtractor.extract` to: `long extract(ConsumerRecord<Object, Object> record, long currentStreamTime)` - this at least gives the user the ability to use the stream time in the event that the time can't be extracted from the record.

mjsax commented:
@miguno You current understanding of the change is no completely correct.

> The default timestamp extractor and the new timestamp extractor differ only in whether or not they are logging a WARN message when a message has an invalid timestamp.

The default TS extractor raises an exception, thus, killing the running instance, while the new TS extractor just logs a WARN message.

I furthermore do understand your argument about splitting extractor and handler. However, from a user point of view, assume a user wants to react on corrupted records and also does provides a custom TS extractor (ie, not using default extractor). First, user must understand that if no valid TS can be extracted, `-1` must be returned, and this will later on trigger a call to an additional error handler she must also provide. In this handler, the handling of the corrupted record will happen. So why not just handle the corrupted record directly in the extractor... that would be way more straight forward.

About local knowledge etc. If Streams policy is to silently drop records with invalid/negative timestamps, the extractor will know whats gonna happening. I don't see a big gab here.

mjsax committed:
addressed github comments
 - added new parameter "Streams time" to extract()
 - added third timestamp extractor

mjsax commented:
Updates this.

Added `InterringConsumerRecordTimestampExtractor` (again, name is subject to discussion).

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:
```diff
@@ -106,7 +106,7 @@ public int addRawRecords(Iterable<ConsumerRecord<byte[], byte[]>> rawRecords, Ti
             log.trace("Source node {} extracted timestamp {} for record {} when adding to buffered queue", source.name(), timestamp, record);
 
-            // validate that timestamp must be non-negative
+            // drop message if TS is invalid, ie, negative
```
Nit:"if timestamp is invalid, i.e., negative"

mjsax replied:
It's not even a JavaDoc comment...

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java:
```diff
@@ -37,8 +37,9 @@
      *
-     * @param record  a data record
+     * @param record a data record
+     * @param currentStreamsTime the current value of the internally tracked Streams time (could be -1 if unknown)
```
"Streams time" -> "stream-time" ?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/InferringConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,74 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Retrieves embedded metadata timestamps from Kafka messages.
```
We need better 1-liners for each extractor. (they are all the same right now)

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/RobustConsumerRecordTimestampExtractor.java:
```diff
@@ -23,43 +23,51 @@
+ * @see InferringConsumerRecordTimestampExtractor
  * @see WallclockTimestampExtractor
  */
 public class RobustConsumerRecordTimestampExtractor implements TimestampExtractor {
```
Name: Perhaps "SkipInvalidTimestampExtractor"?

miguno commented:
Question: The three embedded-timestamp extractors are basically all the same with the exception of the `if (timestamp < 0)` conditional logic.  Should we actually make them have e.g. a common parent class so that users see a proper a type-relationship between them? Or have the default extractor be the base class, and the two new ones override the conditional logic?

mjsax committed:
addressed Github comments

mjsax commented:
@miguno Added `AbstractConsumerRecordTimestampExtractor` as base class for the new extractors.

miguno commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -172,7 +172,7 @@
                                 .define(TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
                                         Type.CLASS,
-                                        ConsumerRecordTimestampExtractor.class.getName(),
+                                        FailingConsumerRecordTimestampExtractor.class.getName(),
```
We should consider what renaming `ConsumerRecordTimestampExtractor` to `FailingConsumerRecordTimestampExtractor` means for backwards compatibility.

mjsax replied:
Because it is the default extractor, use will not explicitly set it. Thus it should not be a big issue.
Of course, strictly it is a breaking change and we must mention it in the KIP (but we need a KIP anyway).

mjsax replied:
We might even keep the name as-is. The reason is, that right now, if `ConsumerRecordTimestampExtractor` returns -1, a failure occurs too. Just not in the extractor but in `RecordQueue`.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/FailingConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @param record a data record
+     * @param recordTimestamp the timestamp extractor from the record
+     * @param currentStreamsTime the current value of the internally tracked Streams time (could be -1 if unknown)
```
Should we be more strict and say: _will_ be -1 if unknown?

miguno replied:
"Streams time" -> "stream-time"

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/InferringConsumerRecordTimestampExtractor.java:
```diff
@@ -19,11 +19,11 @@
 
 /**
  * Retrieves embedded metadata timestamps from Kafka messages.
+ * If a record has a negative (invalid) timestamp value a new timestamp will be inferred as the current stream-time.
```
> If a record has a negative (invalid) timestamp[,] a new timestamp will be inferred [from] the current stream-time.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/SkipInvalidConsumerRecordTimestampExtractor.java:
```diff
@@ -38,36 +41,28 @@
  */
-public class RobustConsumerRecordTimestampExtractor implements TimestampExtractor {
-    private static final Logger log = LoggerFactory.getLogger(RobustConsumerRecordTimestampExtractor.class);
+public class SkipInvalidConsumerRecordTimestampExtractor extends AbstractConsumerRecordTimestampExtractor {
```
What about a shorter name like `SkipRecordsWithInvalidTimestamps`?

Also, I don't think "TimestampExtractor" _must_ be in the name of the class. The term "ConsumerRecord" is also pretty confusing IMHO to the average user because it's not a record OF a consumer etc.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/FailingConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,68 @@
+ * @see InferringConsumerRecordTimestampExtractor
+ * @see WallclockTimestampExtractor
+ */
+public class FailingConsumerRecordTimestampExtractor extends AbstractConsumerRecordTimestampExtractor {
```
If we go with a new name, what about sth like `FailOnRecordsWithInvalidTimestamps`?

My issues with "FailingConsumerRecordTimestampExtractor":
- It sounds as if the extractor itself is failing ("all the time").  That is, the name does not tell you when (and when not) it fails.
- "ConsumerRecord" is IMHO very confusing (see my other comment).
- It is a really, really long name. ;-)
- As I said down below, I don't think "TimestampExtractor" must necessarily be in the name of the implementing classes.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/FailingConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,68 @@
+public class FailingConsumerRecordTimestampExtractor extends AbstractConsumerRecordTimestampExtractor {
+
+    /**
+     * Raised an exception.
```
"Raised an exception"?

I guess: "Raises an exception whenever it is invoked"?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/InferringConsumerRecordTimestampExtractor.java:
```diff
@@ -38,37 +38,33 @@
  * @see WallclockTimestampExtractor
  */
-public class InferringConsumerRecordTimestampExtractor implements TimestampExtractor {
+public class InferringConsumerRecordTimestampExtractor extends AbstractConsumerRecordTimestampExtractor {
```
What about a name like "ReplaceInvalidTimestampWithCurrentStreamTime"?  This suggestion is still a bit too long IMHO, but it is much more descriptive (I think) than "InferringConsumerRecordTimestampExtractor".

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/InferringConsumerRecordTimestampExtractor.java:
```diff
@@ -38,37 +38,33 @@
+                                   final long recordTimestamp,
+                                   final long currentStreamsTime)
+            throws StreamsException {
+        if (currentStreamsTime == -1) {
```
Should we say `== -1` or `< 0`?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/SkipInvalidConsumerRecordTimestampExtractor.java:
```diff
@@ -38,36 +41,28 @@
-     * Extracts the embedded metadata timestamp from the given {@link ConsumerRecord}
-     * and writes a log warn message if the extracted timestamp is negative,
-     * because the record will no be processed but dropped.
+     * Writes a log warn message as the extracted timestamp is negative and the record will no be processed but dropped.
```
Typos:

> Writes a log WARN message when the extracted timestamp is invalid (negative) but returns the the invalid timestamp as-is, which ultimately causes the record to be skipped and not to be processed.

miguno reviewed (commented):
We should also add unit tests for the various new extractors.  If, for example, we say that one extractor will always log a WARN message (on invalids), then we should also validate this behavior with a unit test.  Same for e.g. the extractor that always throws an exception (on invalids).

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TimestampExtractor.java:
```diff
@@ -27,17 +27,19 @@
 
     /**
-     * Extracts a timestamp from a record.
+     * Extracts a timestamp from a record. The timestamp must be positive. If a negative timestamp is returned,
```
I'd suggest:

> Extracts a timestamp from a record.  The timestamp must be positive to be considered a valid timestamp.  Returning a negative timestamp will cause the record not to be processed but rather silently skipped.

Question: What happens when `timestamp == 0`?  Is this a valid or an invalid timestamp?  The answer should get included in the javadoc.

mjsax replied:
Agreed. I would like to write `The timestamp must be non-negative` which includes zero, but this is hard to parse to the user... But `The timestamp must be zero or positive` does also sound clumsy... Not sure how to phrase is nicely... Any suggestion?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/WallclockTimestampExtractor.java:
```diff
@@ -21,22 +21,26 @@
-     * @param record  a data record
-     * @return        the current wall clock time, expressed in milliseconds since midnight, January 1, 1970 UTC
+     * @param record a data record
+     * @param currentStreamsTime the current value of the internally tracked Streams time (could be -1 if unknown)
```
"Streams time" -> "stream-time"

miguno commented:
Further comments.  Sorry for the back and forth, but I think we really need to come up with a good setup/descriptive names in order to not confuse users. :-)

guozhangwang commented:
One meta-comment: as for the name of `currentStreamTime`, it is not actually the `stream time` of the task we talked about before, but rather the `partition time`, and `stream time` is the min over all task's owned partition's time. Currently we are not exposing this `partition time` anywhere to the users, by doing this we are effectively doing this, and hence users need to think about `partition / stream time` in addition to `record timestamp`.

mjsax commented:
@miguno Thanks. I completely agree that the naming must be chosen wisely :) I was never happy with the naming from the beginning on. Thanks for you input.

@guozhangwang Good point. Will think this through and include a discussion in the KIP.

mjsax committed:
updated parameter and class names
added test for new TimestampExtractors
added new metrics

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/AbstractConsumerRecordTimestampExtractor.java:
```diff
@@ -0,0 +1,77 @@
+ * @see InferringConsumerRecordTimestampExtractor
+ * @see WallclockTimestampExtractor
+ */
+public abstract class AbstractConsumerRecordTimestampExtractor implements TimestampExtractor {
```
Make package private
Change Name

mjsax commented:
@guozhangwang @enothereska @dguy @miguno @hjafarpour
Updates this according to our KIP discussions.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/UsePreviousTimeOnInvalidTimestamp.java:
```diff
@@ -0,0 +1,70 @@
+     * @param record a data record
+     * @param recordTimestamp the timestamp extractor from the record
+     * @param previousTimestamp the latest extracted valid timestamp of the current record's partition˙ (could be -1 if unknown)
+     * @return the current stream-time as new timestamp for the record
```
I think @guozhangwang mentioned this already, but i believe this is actually partition time rather than stream-time?

mjsax replied:
Forgot to update the text of the return statement... Thanks!

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/UsePreviousTimeOnInvalidTimestamp.java:
```diff
@@ -0,0 +1,70 @@
+            throws StreamsException {
+        if (previousTimestamp < 0) {
+            throw new StreamsException("Could not infer new timestamp for input record " + record
+                    + " because current internal Streams time in unknown");
```
Same here

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/TimestampExtractorTest.java:
```diff
@@ -0,0 +1,128 @@
+            public void describeMismatch(Object item, Description mismatchDescription) {}
+
+            @Override
+            public void _dont_implement_Matcher___instead_extend_BaseMatcher_() {}
```
Any reason why you didn't take the advice of this method?

mjsax replied:
Did not pay attention... Will fix it.

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordQueueTest.java:
```diff
@@ -136,4 +138,24 @@ public void shouldThrowStreamsExceptionWhenValueDeserializationFails() throws Ex
+        final List<ConsumerRecord<byte[], byte[]>> records = Collections.singletonList(
+                new ConsumerRecord<>("topic", 1, 1, -1L, TimestampType.CREATE_TIME, 0L, 0, 0, recordKey, recordValue));
+
+        queue.addRawRecords(records, new LogAndSkipOnInvalidTimestamp());
```
Not part of this JIRA, but i'm wondering why the `TimestampExtractor` is a method param rather than a field?

mjsax replied:
I think, it is to allow for different extractors for different topics -- even if the user API does not allow to specify different extractors right now, this should be useful if we allow for this at some point. But @guozhangwang should know better.

guozhangwang replied:
We have thought about different extractors per topic, yes, but even with that I think we can keep a field per `RecordQueue`, so that is not a big deal really and we can consider changing it accordingly if necessary.

dguy reviewed (commented):
A couple of minor comments, but LGTM

mjsax committed:
Addressed Damian's comments

mjsax commented:
KIP-93 was accepted.

Call for final review @guozhangwang @enothereska @dguy @miguno

enothereska commented on checkstyle/import-control.xml:
```diff
@@ -27,6 +27,7 @@
   <allow pkg="javax.management" />
   <allow pkg="org.slf4j" />
   <allow pkg="org.junit" />
+  <allow pkg="org.hamcrest" />
```
Is this absolutely needed? cc @ijuma

dguy replied:
IMO - it is useful and i don't see why it should not be allowed.

ijuma replied:
I'm OK with it although @junrao seemed less convinced last time the subject came up.

mjsax replied:
I did talk to @ijuma and @guozhangwang and both told me they are ok with the change. Why no allowing hamcrest for tests?

junrao replied:
As long as the usage is valid, this is fine for me.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/ExtractRecordMetadataTimestamp.java:
```diff
@@ -0,0 +1,77 @@
+     * @param record a data record
+     * @param recordTimestamp the timestamp extractor from the record
+     * @param previousTimestamp the latest extracted valid timestamp of the current record's partition˙ (could be -1 if unknown)
+     * @return a new timestamp for the record
```
Should it be mentioned here that returning a negative timestamp will result in the record being dropped?

mjsax replied:
It's repetitive as this behavior is explained in detail in the interface, but I'll add it. Can't hurt.

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/TimestampExtractorTest.java:
```diff
@@ -0,0 +1,136 @@
+        final long metadataTimestamp = 42;
+
+        final TimestampExtractor[] extractors = new TimestampExtractor[]{
+            new FailOnInvalidTimestamp(),
```
This is a question of taste, but IMO each of these TimestampExtractor implementations should have their own Test class rather than there being a single class covering all of them.

mjsax replied:
`FailOnInvalidTimestamp`, `LogAndSkipOnInvalidTimestamp`, and `UsePreviousTimeOnInvalidTimestamp` extract record metadata timestamp and are thus similar -- I think it's easier to do a single test class to avoid code duplication (c.f. `#extractMetadataTimestamp()`).

guozhangwang replied:
I would second @dguy for this comment: having different classes' test as test functions within a class has a potential risk that we may miss some test case for some classes when adding more later. It is easier to check that for each XX.class we have an XXTest.class.

dguy reviewed (commented):
Couple of minor comments, but overall LGTM

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -951,6 +956,12 @@ public StreamsMetricsImpl(Metrics metrics) {
             this.taskDestructionSensor = metrics.sensor(sensorNamePrefix + ".task-destruction");
             this.taskDestructionSensor.add(metrics.metricName("task-destruction-rate", metricGrpName, "The average per-second number of destructed tasks", metricTags), new Rate(new Count()));
+
+            this.processedRecordsSensor = metrics.sensor(sensorNamePrefix + ".processed-records");
```
nit: I'd suggest renaming to `fetchedRecordsSensor` or `bufferedRecordsSensor` since it is not really processed yet, and similarly for its sensor string, `.buffered-records`);

Also what metrics are you trying to record really? Note that `Count()` will increment the value forever within a single sampling window, or are you trying to use `Rate`?

guozhangwang replied:
Personally I'd suggest we only keep a skip sensor for user to monitor correctness since the the "processed record sensor" is well covered in the throughput sensor already. And for that skip sensor we can use `Rate()`. For example, as in `record-error-rate` in `Sender.class`.

mjsax replied:
I see you point about the naming. However, reporting buffered record is not too helpful for the user IMHO. So not adding this sensor does make sense to me.

`Count()` is internally. Because we assume that only a few record will be dropped, to me it seems most valuable if we report the absolute number of dropped records. Or do you have a strong objection again `Count()`?

guozhangwang replied:
`Count is a {@link SampledStat} that maintains a simple count of what it has seen.` So with this stat, its value will be increased for the window period, then suddenly drops to zero, then start rising again. So it's hard to alert on such a metric, on the other hand `Rate(Count())` will record the average rate per time unit (here second), so users in practice can easily set a threshold for alerting, and even if they want "zero tolerance", setting the threshold to be 0 can still satisfy their needs.

guozhangwang commented:
I have a question about the added sensors, otherwise LGTM.

Could you add the corresponding upgrade section in docs as well? We'd encourage doing code change and doc change in a single PR so that people won't forget about them.

mjsax committed:
Damian's and Guozhang's comments
Added upgrade section to docs

mjsax committed:
removed 'processed' sensor
split test into multiple classes

mjsax commented:
Updated. @guozhangwang @dguy

dguy reviewed (approved):
Thanks @mjsax LGTM

mjsax committed:
Change metric to use Rate()

mjsax commented:
@guozhangwang Updated according to your comment.

guozhangwang commented:
Unit test passed locally.

guozhangwang commented:
Merged to trunk.
